rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,438 @@
+> behaviour does not blur the image sufficiently to obscure its contents. This
+> could lead to thinking that we could blur in our own thumbnail endpoint, but
+> that would be more resource intensive than using BlurHash (I believe). While
+> BlurHash is technically complex to implement, I do think it is the best
+> approach for reasons I will describe below. LQIP could also be resource
+> intensive for the API, but ameliorated with caching.
+
+#### CSS filter
+
+This is the simplest option and only requires adding a `filter: blur(Xrem)`
+style to sensitive results.
+
+This option has the following downsides:
+
+- It is a basic Gaussian blur and may not sufficiently obscure the image if a
+  low number is used or may create an unsettling blob effect if a higher number
+  is used
+- Gaussian blur may not be as aesthetically pleasing as the alternative
+- It requires the image to be downloaded to be blurred by the client
+
+Initial I speculated that client-side Gaussian blurring may have adverse effects
+on low-spec hardware's accessibility of the site. However, I've tested this
+locally by modifying the frontend to blur all images and visiting the local site
+on my PinePhone, on a low-spec quad-core ARM laptop, and on a US$30 Android
+phone. On none of these devices could I perceive a degradation in performance.
+Keep in mind these are all devices that are already very slow! They were able to
+render a full page of image results blurred at 1rem, without issue.
+
+It has one significant upside in that it is supremely simple to implement, does
+not require any new dependencies, and would only require about few lines of
+frontend code rather than lots of frontend and API code to support it.
+Additionally, the image would quickly load after being unblurred by the user
+because it would just require removing the CSS style rather than getting the
+actual image.
+
+#### BlurHash
+
+This is a significantly more complex option that requires both API and frontend
+changes to support. Rather than sending the image to the client, the thumbnail
+URL would send a hashed version of the image that represents the blur. A client
+side library decodes the hash and displays it in place of the thumbnail. It
+would require two new dependencies, a JavaScript BlurHash decoder to render the
+hash in a canvas and an API library to produce the hash from the thumbnail.
+
+It has the following downsides:
+
+- Significantly more complicated than Gaussian blur
+- Requires conditionally calculating the hash on the API side in the thumbnail
+  endpoint
+  - This would be done by using a new query parameter on the thumbnail endpoint
+    to request the blurred hash
+- Can have a performance impact on the thumbnail endpoint
+- Introduces a delay between when the image is unblurred when it is displayed
+  due to needing to download the actual thumbnail
+- Could have different performance issues if decoding the hash is a heavy
```

Comment:
Much of this section discusses the thumbnail endpoint - do we use thumbnails for the single result page or do we load the whole image? If it's the latter, since we're not serving the images themselves, would we need to switch to using the thumbnail for sensitive results?

Reply:
Per @zackkrida: It appears that we hotlink the upstream image in the single result case. I imagine that would impact how we handle navigating to the single result page for sensitive content and what the blurring process is for that, is that correct?

Reply:
You're right, there's some nuance that needs to be teased out here.

The frontend does use the thumbnail on the single results page if navigating to the result from the search page. This allows us to display the result (albeit at lower quality) until we have the full image ready to display. It prevents some weird pop-in behaviours from happening. In that case, we're already using the thumbnail and so we could delay downloading the full image until it is unblurred.

In the case where someone lands directly on the single results page (without navigating from the search page), we do not use the thumbnail. We'd need to have the frontend decide to use the frontend instead of the full image initially for sensitive results and then only load the full image once it's been "unblurred".

This is mostly moot with the CSS filter approach though, as in that case we can always just apply the filter client side, regardless of where the image is coming from. We would only need to consider these nuances if we went with an API-side blurring approach, BlurHash, LQIP, etc.

Reply:
Ah, okay that's a great point! I didn't realize that we could punt on these nuances with the CSS blur approach. That's even further argument for it at this time 🙂